### PR TITLE
feat(mesh): stream chunking wiring in gossip loop (v2 Step 3 part 2)

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

PR #1173 landed Part 1 of Step 3: the self-contained `ChunkAssembler` (receiver-side reassembly algorithm), memory bounds, and `StreamEntry`/`StreamBatch` proto messages. Nothing was wired to the gossip loop yet — the assembler was dead code, the proto had a `StreamBatch` variant on `StreamMessage.payload` but no `StreamMessageType` discriminator, and no path existed for stream values to actually ride the bidirectional `SyncStream` between peers.

This PR wires it end-to-end: sender-side chunking, receiver-side routing, subscriber fan-out. After this merges, any value published through a `StreamNamespace` buffer or emitted via `register_drain` flows through the mesh to every connected peer's subscribers.

### Solution

Five focused commits:

1. **Extend `StreamMessageType`** with `STREAM_BATCH = 7` and add stub match arms in both `controller.rs` and `ping_server.rs`. Closes Codex P1 from PR #1173.
2. **Plumb `Arc<MeshKV>`** from `MeshServerBuilder` into both `MeshController` and `GossipService` via `with_mesh_kv()`. `ChunkAssembler` now lives on `MeshKV` alongside the subscriber registry — they travel together.
3. **`chunking.rs` helper module**: `chunk_value(key, generation, Bytes, max_chunk_bytes)` splits oversized values into `StreamEntry`s (single-chunk fast path for values ≤ 10 MB); `build_stream_batches` packs and caps at `max_chunks_per_round`. 12 unit tests covering split boundaries, empty/zero edge cases, tree-page invariant. `#[debug_assert]` catches any `tree:page:*` entry that would go multi-chunk (spec §4.1 invariant).
4. **Sender integration**: `MeshController` collects the stream round-batch alongside the CRDT batch once per second. Both sender tasks (controller.rs client-side, ping_server.rs server-side) read the shared `current_stream_batch`, chunk via `chunk_value`, and emit `StreamBatch` messages. Client-side handles both broadcast and targeted entries; server-side handles broadcast only (peer_name isn't known at server-task spawn). Generation is wall-clock nanoseconds (monotonic across restarts). At-most-once per §4.4: on channel full, the round's stream traffic is dropped — no NACK, no retry.
5. **Receiver integration + fragmented-buffer API**: new `dispatch_stream_batch` helper handles both fast-path (total_chunks=1) and chunked entries. Fires subscribers via `MeshKV::notify_subscribers`. Periodic chunk-assembler GC from the main event loop (every 5s, 30s timeout). Subscriber API changes to `Option<Vec<Bytes>>` so zero-copy fan-out survives all the way to adapter channels — closes the Codex-deferred "2× payload during assembly" concern from PR #1173 (receiver now peaks at P bytes, not 2P, because no contiguous concat happens).

## Changes

- `crates/mesh/src/proto/gossip.proto`: `StreamMessageType::STREAM_BATCH` variant.
- `crates/mesh/src/chunking.rs`: new module with `chunk_value`, `build_stream_batches`, `dispatch_stream_batch` + 12 unit tests.
- `crates/mesh/src/chunk_assembler.rs`: `assemble()` now returns `Vec<Bytes>` (zero-copy fragmented). `receive_chunk` signature updated. Dropped the `#![allow(dead_code)]`.
- `crates/mesh/src/kv.rs`: `Arc<ChunkAssembler>` field on `MeshKV`. `notify_subscribers(&str, Option<Vec<Bytes>>)` accessor. `SubscriptionEvent` / `SubscriberRegistry::notify` signatures updated to fragmented buffers. `CrdtNamespace::put` wraps single local writes in a 1-element Vec. `RoundBatch` derives `Default`. Manual `Debug` impl for `MeshKV`.
- `crates/mesh/src/controller.rs`: `current_stream_batch` field + accessor, per-round collection in event loop, per-peer sender emits `StreamBatch` (broadcast + targeted), receiver fills `StreamMessageType::StreamBatch` arm, periodic chunk-assembler GC.
- `crates/mesh/src/ping_server.rs`: `current_stream_batch` + `mesh_kv` fields, `with_current_stream_batch` / `with_mesh_kv` builders, server-side sender emits broadcast stream batches, receiver fills `StreamMessageType::StreamBatch` arm.
- `crates/mesh/src/service.rs`: creates `Arc<MeshKV>` in `build()` and plumbs into both controller and service.
- `crates/mesh/src/lib.rs`: `mod chunking`.

## Test Plan

Before this PR: stream values had no transport path; any `publish_to` or drain-callback output sat in in-memory buffers and never reached other nodes.

After this PR:

```
$ cargo test -p smg-mesh --lib
    test result: ok. 230 passed; 0 failed; 2 ignored

$ cargo clippy -p smg-mesh --all-targets --all-features -- -D warnings
    Finished dev [unoptimized + debuginfo]

$ cargo +nightly fmt
    (clean)
```

Coverage:
- **chunk_value**: single-chunk fast path, empty, even/uneven split boundaries, exact-boundary, one-byte-over, zero-max-panic, tree-page fast-path, tree-page debug-assert panic on multi-chunk.
- **build_stream_batches**: empty, under-cap, over-cap truncation.
- **ChunkAssembler** (now returning Vec<Bytes>): all prior tests updated with a flatten test-helper; 18 tests pass.
- **kv subscribe**: fragment-aware assertion on single-element Vec for local writes.

Integration tests for multi-node round-trip, partial-delivery timeout, concurrent generations, and multi-model concurrent assembly land in follow-up commits (Step 3.6 + 3.7) on a fresh branch after this merges — those touch `tests/comprehensive.rs` and don't overlap with the files changed here.

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [x] (Optional) Documentation updated — inline module docs on chunking.rs; chunk_assembler.rs public contract note on fragmented return
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

## Summary by CodeRabbit

* **New Features**
  * Chunked streaming for large values with batch packing, dispatch, and automatic reassembly.
  * Controller/server now coordinate and emit stream batches during rounds; peers receive and apply batches, notifying subscribers only on full reassembly.

* **Refactor**
  * Subscription events and messaging use fragmented payloads (vectors of fragments) for delivery.
  * In-flight assemblies are tracked per-sender to isolate concurrent streams and avoid cross-peer collisions.

### CODE REVIEW

File: crates/mesh/src/chunking.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+/// already drained and the application will re-publish on its next
+/// retry cycle.
+pub fn build_stream_batches(
+    entries: Vec<StreamEntry>,
+    max_chunks_per_round: usize,
+) -> Vec<StreamBatch> {
+    if entries.is_empty() {
+        return Vec::new();
+    }
+    let taken: Vec<StreamEntry> = entries.into_iter().take(max_chunks_per_round).collect();
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in 1f4db260 — build_stream_batches now packs into multiple StreamBatches of at most max_chunks_per_batch each instead of truncating. Every entry is emitted; backpressure on tx.try_send still provides per-round rate limiting. Renamed parameter to max_chunks_per_batch to match the new semantics.

File: crates/mesh/src/chunking.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    if taken.is_empty() {
+        return Vec::new();
+    }
+    vec![StreamBatch { entries: taken }]
+}
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs:200]: Fixed in 43d3150a — added debug_assert!(max_chunks_per_batch > 0, ...). Release-mode runtime zero-check stays as a safety net. Test updated to should_panic under debug_assertions.

File: crates/mesh/src/chunking.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+/// Receiver-side dispatch for `StreamBatch` entries. Single-chunk
+/// entries (`total_chunks == 1`) fire subscribers directly — no state
+/// in the chunk assembler. Multi-chunk entries route through the
+/// assembler and fire subscribers only on full reassembly. Fires go
+/// through `MeshKV::notify_subscribers` with a fragmented `Vec<Bytes>`
+/// payload so fan-out is zero-copy.
+pub fn dispatch_stream_batch<'a>(
+    mesh_kv: &MeshKV,
+    entries: impl IntoIterator<Item = &'a StreamEntry>,
+) {
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in 2d0f533f — dispatch_stream_batch now takes IntoIterator<Item = StreamEntry> (owned). Both controller.rs and ping_server.rs destructure msg.payload by value instead of by reference so batch.entries moves in, making Bytes::from(entry.data) zero-copy.

File: crates/mesh/src/chunking.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    for entry in entries {
+        if entry.total_chunks == 1 {
+            mesh_kv.notify_subscribers(&entry.key, Some(vec![Bytes::from(entry.data.clone())]));
+        } else if let Some(fragments) = mesh_kv.chunk_assembler().receive_chunk(
+            &entry.key,
+            entry.generation,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in e7393745 — added `peer_id` to `ChunkAssembler::receive_chunk` and dispatch_stream_batch, internally scoping `AssemblyState` by `format!("{peer_id}\\0{key}")`. '\\0' can't appear in `peer_id` or subscriber keys so the composite key doesn't collide with any unscoped application key. Gossip call sites pass `&msg.peer_id`.

File: crates/mesh/src/chunking.rs [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    key: String,
+    generation: u64,
+    value: Bytes,
+    max_chunk_bytes: usize,
+) -> Vec<StreamEntry> {
+    assert!(max_chunk_bytes > 0, "max_chunk_bytes must be non-zero");
+
+    if value.len() <= max_chunk_bytes {
+        debug_assert!(
+            !key.starts_with("tree:page:") || value.len() <= max_chunk_bytes,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in 006bbca7 — tautological debug_assert removed as part of a broader doc trim pass on chunking.rs.

File: crates/mesh/src/chunking.rs [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            "tree:page:* entries must fit in a single gRPC message"
+        );
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in 006bbca7 — removed; see also the module-level doc refresh in the same commit.

File: crates/mesh/src/kv.rs [hunk 7]
```diff
@@ -463,11 +483,27 @@ impl MeshKV {
...
+        self.chunk_assembler.clone()
+    }
+
+    /// Fire subscribers whose prefix matches `key`. Used by the gossip
+    /// receive path when a chunked value completes (or a single-chunk
+    /// entry arrives), so handlers can deliver into adapter-owned
+    /// mpsc channels without reaching into internal registries.
+    pub fn notify_subscribers(&self, key: &str, value: Option<Vec<Bytes>>) {
+        self.subscriber_registry.notify(key, value);
+    }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/kv.rs:505]: Fixed in c599a106 — both are now pub(crate). chunking.rs (their only caller) is also crate-internal so nothing breaks.

File: crates/mesh/src/ping_server.rs [hunk 8]
```diff
@@ -529,6 +567,60 @@ impl Gossip for GossipService {
...
+                    // (controller.rs) where peer_name is known at task
+                    // spawn. At-most-once: on channel full, drop without
+                    // retry (spec §4.4).
+                    if let Some(sbh) = &stream_batch_handle {
+                        let stream_batch = sbh.read().clone();
+                        let fresh_batch = last_stream_batch
+                            .as_ref()
+                            .is_none_or(|last| !Arc::ptr_eq(last, &stream_batch));
+                        if fresh_batch && !stream_batch.drain_entries.is_empty() {
+                            last_stream_batch = Some(stream_batch.clone());
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/ping_server.rs]: Fixed in 2361d0bc. `last_stream_batch` now advances on every fresh Arc, not only when `drain_entries` is non-empty, so an initially-empty batch stops looking "fresh" every tick.

File: crates/mesh/src/controller.rs [hunk 9]
```diff
@@ -592,6 +651,69 @@ impl MeshController {
...
+                                    .map(|d| d.as_nanos() as u64)
+                                    .unwrap_or(0);
+                                let mut entries = Vec::new();
+                                // Drain entries are broadcast: every peer emits.
+                                for (key, value) in &stream_batch.drain_entries {
+                                    entries.extend(chunk_value(
+                                        key.clone(),
+                                        generation,
+                                        Bytes::from(value.clone()),
+                                        MAX_MESSAGE_SIZE,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Fixed in 1a77f372. Introduced `MAX_STREAM_CHUNK_BYTES = MAX_MESSAGE_SIZE - 64 KiB` (`STREAM_CHUNK_OVERHEAD_MARGIN`) and chunk_value now uses it, leaving headroom for the protobuf envelope (StreamMessage/StreamEntry metadata + tags).

File: crates/mesh/src/ping_server.rs [hunk 10]
```diff
@@ -529,6 +567,60 @@ impl Gossip for GossipService {
...
+                                .duration_since(std::time::UNIX_EPOCH)
+                                .map(|d| d.as_nanos() as u64)
+                                .unwrap_or(0);
+                            let mut entries = Vec::new();
+                            for (key, value) in &stream_batch.drain_entries {
+                                entries.extend(chunk_value(
+                                    key.clone(),
+                                    generation,
+                                    Bytes::from(value.clone()),
+                                    MAX_MESSAGE_SIZE,
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/ping_server.rs]: Fixed in 1a77f372 — same fix as the client-side site, ping_server.rs now passes `MAX_STREAM_CHUNK_BYTES` which leaves 64 KiB of gRPC envelope headroom below `MAX_MESSAGE_SIZE`.

File: crates/mesh/src/chunking.rs [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+use bytes::Bytes;
+
+use crate::{
+    kv::MeshKV,
+    service::gossip::{StreamBatch, StreamEntry},
+};
+
+/// Per-round cap on chunks emitted across all stream entries. Prevents
+/// a burst of large values from monopolising the gossip channel.
+pub const DEFAULT_MAX_CHUNKS_PER_ROUND: usize = 5;
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in 8e9701aa (renamed to `DEFAULT_MAX_CHUNKS_PER_BATCH`) and 1a77f372 + e331754a (the byte cap). Doc comment now states explicitly that there is no per-round bandwidth cap — multiple batches are emitted per round if needed, since a hard per-round cap combined with §4.4 at-most-once would strand oversized values undeliverable.

File: crates/mesh/src/controller.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Fixed in 1a77f372. Added `STREAM_CHUNK_OVERHEAD_MARGIN = 64 KiB` and `MAX_STREAM_CHUNK_BYTES = MAX_MESSAGE_SIZE - margin`. Chunk bodies now leave room for the protobuf envelope so the serialised StreamMessage stays under the tonic decoding cap.

File: crates/mesh/src/controller.rs [hunk 12]
```diff
@@ -592,6 +651,69 @@ impl MeshController {
...
+                                            peer_id: self_name_incremental.clone(),
+                                        };
+                                        if let Err(e) = tx_incremental.try_send(msg) {
+                                            log::debug!(
+                                                peer = %peer_name_incremental,
+                                                error = ?e,
+                                                "stream batch dropped on backpressure"
+                                            );
+                                            // TODO(metrics): bump stream_dropped_on_backpressure
+                                            break;
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Fixed in 454fffa8. The `try_send` match now distinguishes `TrySendError::Full` (backpressure → break this round) from `TrySendError::Closed` (receiver gone → `return` out of the sender task). Applied to both client-side (controller.rs) and server-side (ping_server.rs).

File: crates/mesh/src/service.rs [hunk 13]
```diff
@@ -333,6 +334,7 @@ impl MeshServerBuilder {
                 signal_rx,
                 partition_detector: Some(partition_detector.clone()),
                 mtls_manager: self.mtls_manager.clone(),
+                mesh_kv,
             },
             MeshServerHandler {
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/service.rs]: Fixed in 06fafeb6. MeshServerHandler now holds the same `Arc<MeshKV>` as MeshServer and exposes it via `handler.mesh_kv()` so adapters can subscribe and publish through a single shared instance.

File: crates/mesh/src/controller.rs [hunk 14]
```diff
@@ -592,6 +651,69 @@ impl MeshController {
...
+                            let stream_batch = stream_batch_handle.read().clone();
+                            let fresh_batch = last_stream_batch
+                                .as_ref()
+                                .is_none_or(|last| !Arc::ptr_eq(last, &stream_batch));
+                            if fresh_batch {
+                                last_stream_batch = Some(stream_batch.clone());
+                                let generation = std::time::SystemTime::now()
+                                    .duration_since(std::time::UNIX_EPOCH)
+                                    .map(|d| d.as_nanos() as u64)
+                                    .unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Fixed in 57403de9. Replaced the per-round wall-clock timestamp with a process-wide `LazyLock<AtomicU64>` seeded from wall-clock nanos. Restarts re-seed from the clock (so fresh generations almost certainly exceed pre-restart ones), but mid-process increments are pure atomics, so NTP steps can't rewind the counter. `next_generation()` is called per value.

File: crates/mesh/src/controller.rs [hunk 15]
```diff
@@ -592,6 +652,69 @@ impl MeshController {
...
+                                }
+                                // Targeted entries: only those addressed to this peer.
+                                for (target, key, value) in &stream_batch.targeted_entries {
+                                    if target == &peer_name_incremental {
+                                        entries.extend(chunk_value(
+                                            key.clone(),
+                                            generation,
+                                            value.clone(),
+                                            MAX_STREAM_CHUNK_BYTES,
+                                        ));
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Fixed in 57403de9 (same as r3112439424 / r3112027060). Generation is now produced by `next_generation()` per value, not per round — each published value gets a fresh monotonic tag. Backing store is a `LazyLock<AtomicU64>` seeded from wall-clock nanos so restart-freshness survives while mid-process timer jumps can't rewind it.

File: crates/mesh/src/chunking.rs [hunk 16]
```diff
@@ -0,0 +1,292 @@
...
+/// the same key don't collide.
+pub fn dispatch_stream_batch(
+    mesh_kv: &MeshKV,
+    peer_id: &str,
+    entries: impl IntoIterator<Item = StreamEntry>,
+) {
+    for entry in entries {
+        if entry.total_chunks == 1 {
+            mesh_kv.notify_subscribers(&entry.key, Some(vec![Bytes::from(entry.data)]));
+        } else {
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs:138]: Fixed in 4c5cefb1. Added `ChunkAssembler::drop_pending(peer_id, key)`; the single-chunk fast path in `dispatch_stream_batch` calls it before notifying subscribers so a fresh single-chunk value supersedes any lingering fragments from an older multi-chunk publish under the same (peer, key).

File: crates/mesh/src/chunking.rs [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+        "max_chunks_per_batch must be non-zero; callers should pass \
+         DEFAULT_MAX_CHUNKS_PER_ROUND or another positive cap"
+    );
+    if max_chunks_per_batch == 0 || entries.is_empty() {
+        return Vec::new();
+    }
+    let mut out = Vec::new();
+    let mut iter = entries.into_iter().peekable();
+    while iter.peek().is_some() {
+        let batch: Vec<StreamEntry> = iter.by_ref().take(max_chunks_per_batch).collect();
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/chunking.rs]: Fixed in e331754a. `build_stream_batches` now takes `max_bytes_per_batch` and flushes whenever adding the next entry would exceed either cap. Callers pass `MAX_STREAM_CHUNK_BYTES` so the whole StreamBatch fits in one gRPC message. An entry that on its own exceeds the byte cap still ships as a single-entry batch — dropping would strand downstream chunks undeliverable under at-most-once.

File: crates/mesh/src/ping_server.rs [hunk 18]
```diff
@@ -529,6 +567,60 @@ impl Gossip for GossipService {
...
+                                ));
+                            }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/ping_server.rs]: Fixed in 57403de9. `next_generation()` in chunking.rs uses a `LazyLock<AtomicU64>` seeded from wall-clock nanos on first use; every published value gets a fresh tag (per-value, not per-round/per-batch).

File: crates/mesh/src/controller.rs [hunk 19]
```diff
@@ -1011,6 +1133,15 @@ impl MeshController {
...
                                 }
+                                StreamMessageType::StreamBatch => {
+                                    if let (
+                                        Some(StreamPayload::StreamBatch(batch)),
+                                        Some(mesh_kv),
+                                    ) = (&msg.payload, &mesh_kv)
+                                    {
+                                        dispatch_stream_batch(mesh_kv, batch.entries.iter());
+                                    }
+                                }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs]: Addressed in 2d0f533f (Fix P): `dispatch_stream_batch` now takes owned `StreamEntry`s and moves the payload into `Bytes` without cloning — the receiver arm matches the payload once and hands ownership through. No additional clone between receive and subscriber fan-out.

File: crates/mesh/src/ping_server.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/ping_server.rs]: Tracked for follow-up. Server-side targeted routing needs the learned `peer_id` plumbed through to the emission loop and the targeting check itself is non-trivial (client-side has `peer_name` at task spawn; server side has to wait until the inbound side identifies the peer). Keeping the current behaviour — broadcast drains server-side, targeted client-side — for this PR; opening a dedicated follow-up issue so we can land it with its own tests.

File: crates/mesh/src/controller.rs [hunk 20]
```diff
@@ -223,6 +267,16 @@ impl MeshController {
...
             }
 
+            // Stream round collection: drain stream namespace buffers and
+            // drain callbacks exactly once per round (destructive). Per-peer
+            // senders filter targeted_entries by their own peer_id and
+            // broadcast drain_entries to all peers. Empty batch if no
+            // MeshKV is attached (legacy path pre-Step 3).
+            if let Some(mesh_kv) = &self.mesh_kv {
+                let stream_batch = mesh_kv.collect_round_batch();
+                *self.current_stream_batch.write() = Arc::new(stream_batch);
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/controller.rs:277]: Tracked for follow-up (separate PR). Switching from replace-with-Arc to a queue-of-rounds model is a design change (who owns the queue, how many rounds of backlog to keep, interaction with §4.4 at-most-once) and we want to land it with its own design review + tests rather than fold it into the chunking transport PR. Current drop-on-backpressure behaviour matches spec §4.4 in the interim.

File: crates/mesh/src/ping_server.rs [hunk 21]
```diff
@@ -1242,6 +1336,15 @@ impl Gossip for GossipService {
...
                             }
+                            StreamMessageType::StreamBatch => {
+                                if let (
+                                    Some(gossip::stream_message::Payload::StreamBatch(batch)),
+                                    Some(mesh_kv),
+                                ) = (&msg.payload, &mesh_kv)
+                                {
+                                    dispatch_stream_batch(mesh_kv, batch.entries.iter());
+                                }
+                            }
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/mesh/src/ping_server.rs:1364]: Addressed in 2d0f533f (Fix P): `dispatch_stream_batch(mesh_kv, peer_id, batch.entries)` takes owned entries, so the server-side receiver hands the `Vec<StreamEntry>` over without cloning. Payloads move into `Bytes` for zero-copy fan-out to subscribers.
